### ISSUE TITLE
Share Utilization Soft Limit?

### ISSUE THREAD

**Issue by baggar11** [rockstor/rockstor-core#782]
I ran through a couple tests where I created 3 types of pools with zlib compression. 1 disk as single, 2 disk as raid1 and a 3 disk as raid5. I shared those out via Samba and started pushing data over to them. Each one of them failed to fill up at around 60-65% of share utilization. All of them gave "out of space" messages.

**Comment by schakrava** [comment]
@baggar11 I am working on this, but have a question for you. How big are your disks? And optionally, is the behavior same without any compression?

**Comment by baggar11** [comment]
I believe it happens without, I'll run through that again. It will take me a couple days though.

My disks are as follows.

500gb (boot)
2tb
400gb
300gb
250gb
150gb
2 x 80gb
70gb
40gb

**Comment by schakrava** [comment]
No worries. This is a known issue(being worked on -- #780), but just want to get more specific data for... science!

**Comment by baggar11** [comment]
I'll have to rebuild the box since I'm currently using it for testing out all kinds of stuff at the moment, but it went like this.

2TB pool1(single disk) - share1(full utilization)
2 x 80gb pool2(raid1) - share2(full utilization)
400gb + 300gb + 250gb pool3(raid5) - share3(full utilization)

**Comment by baggar11** [comment]
Ok, finally got around to rebuilding my test machine with RockStor after trying to virtualize the testing. Interestingly though, RockStor would hard lock without host errors when trying to virtualize via Proxmox.

Anyways, running the same copying test without compression, here are the results. As you can see, all of the shares started giving out of space messages at different utilization levels. Hope that helps.

//192.168.10.104/movies       70G   47G   23G  68% /home/baggar11/rockstor_movies
//192.168.10.104/pictures     75G   41G   34G  55% /home/baggar11/rockstor_pictures
//192.168.10.104/tvshows     680G  244G  312G  44% /home/baggar11/rockstor_tvshows

**Comment by phillxnet** [comment]
@baggar11 Thanks for your efforts with this, I believe @schakrava is currently investigating this issue further and it looks like it might well be btrfs / qgroups related and not Rockstor specific which is both good and bad. Watch this space for updates.

**Comment by schakrava** [comment]
I got the confirmation from the qgroup developer(at least one of them anyway) that support is much better in 4.2. So for now, I've made the change that's referenced above to temporarily disable Share size enforcement and display appropriate warnings in the UI.

4.2 should be released sometime this month and we'll get going with it in our testing repo and hopefully we can enable this feature back in a month or so.

**Comment by baggar11** [comment]
Did this get merged into the 3.8-6 update? I noticed the message about share size enforcement is temporarily disabled now. So I ran through some fresh tests and am still having issues around 74%.

**Comment by schakrava** [comment]
@baggar11 yes, quota enforcement is disabled for now. Can you provide more information about your tests. Are you using Shares from before the update? If so, try running /opt/rockstor/bin/qgroup-maxout-limit. I'd be interested in what the output of this script looks like.

**Comment by baggar11** [comment]
These are all new pools and shares. I've noticed that I don't get any screen graphs on the dashboard for CPU and Network. Maybe my install is borked after some of my previous testing?

Here's the output of qgroup-maxout-limit
-----------------------------------------------------

[root@rockstor ~]# /opt/rockstor/bin/qgroup-maxout-limit 
Processing pool(pool2-raid5)
No parent qgroup for 0/5          16.00KiB     16.00KiB ---     
No parent qgroup for 2015/1      379.83GiB    379.83GiB ---     
No parent qgroup for 2015/2      240.00KiB    240.00KiB ---     
Ignoring unexcepted line().
relaxing the limit on qgroup 2015/1
relaxing the limit on qgroup 2015/2
Finished processing pool(pool2-raid5)
Processing pool(rockstor_rockstor)
No parent qgroup for 0/5          16.00KiB     16.00KiB ---     
No parent qgroup for 0/257        16.00KiB     16.00KiB ---     
No parent qgroup for 0/258         1.62GiB      1.62GiB ---     
No parent qgroup for 2015/1          0.00B        0.00B ---     
No parent qgroup for 2015/2          0.00B        0.00B ---     
Ignoring unexcepted line().
Finished processing pool(rockstor_rockstor)
Processing pool(pool1-single)
No parent qgroup for 0/5          16.00KiB     16.00KiB ---     
No parent qgroup for 2015/1       59.50GiB     59.50GiB ---     
No parent qgroup for 2015/2       16.00KiB     16.00KiB ---     
Ignoring unexcepted line().
relaxing the limit on qgroup 2015/1
relaxing the limit on qgroup 2015/2
Finished processing pool(pool1-single)

**Comment by baggar11** [comment]
Ok, just got done pumping data over to a new install of RockStor with fresh pools/shares. Still cuts out at 74% utilization. Here's the latest qgroup-maxout-limit printout. Hope it helps.

/opt/rockstor/bin/qgroup-maxout-limit
Processing pool(rockstor_rockstor)
No parent qgroup for 0/5          16.00KiB     16.00KiB ---
No parent qgroup for 0/257        16.00KiB     16.00KiB ---
No parent qgroup for 0/258         1.63GiB      1.63GiB ---
No parent qgroup for 2015/1          0.00B        0.00B ---
No parent qgroup for 2015/2          0.00B        0.00B ---
Ignoring unexcepted line().
Finished processing pool(rockstor_rockstor)
Processing pool(pool1-single)
No parent qgroup for 0/5          16.00KiB     16.00KiB ---
No parent qgroup for 2015/1       59.68GiB     59.68GiB ---
No parent qgroup for 2015/2       87.27GiB     87.27GiB ---
Ignoring unexcepted line().
relaxing the limit on qgroup 2015/1
relaxing the limit on qgroup 2015/2
Finished processing pool(pool1-single)
Processing pool(pool2-raid5)
No parent qgroup for 0/5          16.00KiB     16.00KiB ---
No parent qgroup for 2015/1      379.83GiB    379.83GiB ---
No parent qgroup for 2015/2       16.00KiB     16.00KiB ---
Ignoring unexcepted line().
relaxing the limit on qgroup 2015/1
relaxing the limit on qgroup 2015/2
Finished processing pool(pool2-raid5)

**Comment by baggar11** [comment]
Ok, I wiped the drives again, reinstalled RockStor and performed the test again. Different result this time.

pool1-single is a single disk with no compression. I've got 2 cifs shares attached. I copied data over until they filled up to 100%. So no issues there.

pool2-radid5 is a mix of 3 disks with zlib compression turned on. Copying data over, I can get to 74% before the transfers die.

Here is where I think the problem lies. I think the total available space for the pool is being miscalculated. disk1 = 372.60g, disk2 = 232.90g, disk3 = 149g. RockStor reports the pool as 572.97g. CIFS reports it as 755g. When I put those disks into the btrfs calculator here --> http://carfax.org.uk/btrfs-usage/ it tells me that the pool size should be 381g. Which is exactly where my transfers die off at.

I hope this helps.

**Comment by schakrava** [comment]
Definitely helps. Should be an easy fix. Thanks @baggar11

**Comment by schakrava** [comment]
Ok, I have news about this issue. Looks like btrfs qgroup stuff is working now. I've tested it on 4.3.3 and also added bin/qgroup-test test script that tests the behavior. Still not comfortable pushing the changes to Testing, but definitely a great sign.

**Comment by baggar11** [comment]
Great news! Any idea on how you might attack how Samba is reporting total volume sizes?

**Comment by phillxnet** [comment]
Thanks to forum member coolbit for bringing share size enforcement to the fore again.

Having now pretty much completed our transition to an openSUSE base with our Rockstor 4 'Built on openSUSE' variant/effort, where we inherit a Leap 15.2 upstream supported/updated kernel, I think we are ready to again look to re-enabling share size quota enforcement.

Our base kernel version, in Rockstor 4, is now 5.3 with aggressive and ongoing openSUSE orchestrated btrfs backports. So we should be in a better position to re-enable this core capability re share size enforcement via quotas.

Referenced forum thread: https://forum.rockstor.com/t/problems-shares/7309
